pkg/teeth/github.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package teeth
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/abcxyz/github-metrics-aggregator/pkg/githubclient"
+	"github.com/google/go-github/v56/github"
+	"github.com/sethvargo/go-retry"
+)
+
+const ()
+
+var (
+	// can be overriden for testing
+	retryMinWaitDuration        = 1 * time.Second
+	retryMaxAttempts     uint64 = 4
+	retryFunc                   = retry.NewFibonacci
+)
+
+// GHConfig contains configuration parameters for the GitHub client.
+type GHConfig struct {
+	Owner string
+	Repo  string
+}
+
+// GitHub is just a wrapper around githubclient.GitHub for functionality specific to teeth.
+// Use NewGitHub to initialize this.
+type GitHub struct {
+	*githubclient.GitHub
+
+	config *GHConfig
+	// can be overriden for testing
+	commentPRFunc func(ctx context.Context, owner, repo string, prNumber int, comment string) (*github.Response, error)
+}
+
+func NewGitHub(ctx context.Context, config *GHConfig, appID, rsaPrivateKeyPEM string) (*GitHub, error) {
+	client, err := githubclient.New(ctx, appID, rsaPrivateKeyPEM)
+	if err != nil {
+		return nil, fmt.Errorf("failed to setup github client: %w", err)
+	}
+	return &GitHub{
+		GitHub:        client,
+		config:        config,
+		commentPRFunc: client.CommentPR,
+	}, nil
+}
+
+func convertRetryable(statusCode int) error {
```

Comment:
Naming nit: `shouldRetry`? It feels weird that this returns an error instead of a boolean too.

Reply:
Feels like this should maybe live in `pkg/githubclient`

Reply:
> Naming nit: shouldRetry? It feels weird that this returns an error instead of a boolean too.
Done. Also agree that `shouldRetry(...) bool` is more customary .

> Feels like this should maybe live in pkg/githubclient
Good point. I thought about it but pkg/githubclient has some other logic for the GMA retry service as well and I'm not familiar with the calls it makes. I lean toward giving the caller responsibility on which statuses to retry based on the specific Github API and scoping the githubclient to just handle requests and responses.